libobs: Add wayland hotkey infrastructure

kkartaltepe opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Make sure you’ve read the contribution guidelines here: https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst -->

### Description
<!--- Describe your changes in detail. -->
<!--- If this change includes UI elements, please include screenshots. -->
Users on wayland are displeased that they cannot see their hotkey
bindings. This enables keys reported like X11, and has the infrastructure
in place in case wayland ever decides to allow for capturing input.

Also fix numpad subtract missing on X11.

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open GitHub Issue, or implements feature request -->
<!--- from the Ideas page, please link to the issue here. -->
Users cannot even see or set hotkeys which might be usable while OBS is in focus. Hotkeys now work while OBS is in focus and display properly on the UI.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment (hardware, OS version, etc.),-->
<!--- and the tests you ran, including how it may affect other areas of code. -->
Tested in kde that keymapping tables work correctly on US layouts and the keys on my keyboard that work in X11 work in Wayland.

More testing on non-US layouts would be appreciated.

### Types of changes
<!--- What types of changes does your PR introduce? Uncomment all that apply -->
 - Bug fix (non-breaking change which fixes an issue) 
<!--- - New feature (non-breaking change which adds functionality) -->
<!--- - Tweak (non-breaking change to improve existing functionality) -->
<!--- - Performance enhancement (non-breaking change which improves efficiency) -->
<!--- - Code cleanup (non-breaking change which makes code smaller or more readable) -->
<!--- - Breaking change (fix or feature that would cause existing functionality to change) -->
<!--- - Documentation (a change to documentation pages) -->

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

kkartaltepe commented:
Also most of the key translation was done programmatically, some keys may be missing the only ones I added manually were numpad keys.

WizardCM commented on libobs/obs-nix-x11.c:
```diff
@@ -1179,6 +1179,8 @@ static void obs_nix_x11_key_to_str(obs_key_t key, struct dstr *dstr)
 		return translate_key(key, "Numpad *");
 	case OBS_KEY_NUMPLUS:
 		return translate_key(key, "Numpad +");
+	case OBS_KEY_NUMMINUS:
+		return translate_key(key, "Numpad -");
```
This should likely be a separate commit.

kkartaltepe replied:
Sounds good.

GeorgesStavracas commented on libobs/obs-nix-wayland.c:
```diff
@@ -19,7 +19,224 @@
+	// My setup doesnt have multiple layouts, do people actually use these?
+	// int num_layouts = xkb_keymap_num_layouts_for_key(keymap, key);
+	// blog(LOG_INFO, "key layouts: %d", num_layouts);
+	for (int l = 0; l < MAX_SHIFT_LEVELS; l++) {
```
`int level` perhaps?

GeorgesStavracas commented on libobs/obs-nix-wayland.c:
```diff
@@ -19,7 +19,224 @@
+// X11 only supports 256 scancodes, most keyboards dont have 256 keys so this should be reasonable.
+#define MAX_KEYCODES 256
+// X11 keymaps only have 4 shift levels, im not sure xkbcommon supports a way to shift the state into a higher level anyway.
+#define MAX_SHIFT_LEVELS 4
```
Can this number be derived from the current keymap?

kkartaltepe replied:
These could be derived but if we consider non-X11 compatible both keycodes and shift levels have no limit. So such a simple implementation may not be feasible. Only supporting X11 compatible keymaps seems like a nice middle ground of simplicity and supporting all existing (x11 compatible) layouts.

GeorgesStavracas commented on libobs/obs-nix-wayland.c:
```diff
@@ -19,7 +19,224 @@
+#include <xkbcommon/xkbcommon.h>
+
+// X11 only supports 256 scancodes, most keyboards dont have 256 keys so this should be reasonable.
+#define MAX_KEYCODES 256
```
Same for this one

GeorgesStavracas commented on libobs/obs-nix-wayland.c:
```diff
@@ -19,7 +19,224 @@
+
+	char *keymap_shm = mmap(NULL, size, PROT_READ, MAP_SHARED, fd, 0);
+	if (keymap_shm == MAP_FAILED) {
+		return;
```
I'm not sure if it's necessary to `close(fd)` here as well

GeorgesStavracas commented on libobs/obs-nix-wayland.c:
```diff
@@ -51,24 +281,1337 @@ obs_nix_wayland_hotkeys_platform_is_pressed(obs_hotkeys_platform_t *context,
+		return;
+	}
+
+	/*
+	obs_hotkeys_platform_t *context = obs->hotkeys.platform_context;
+	struct keycode_list *keycodes = &context->keycodes[key];
+
+	for (size_t i = 0; i < keycodes->list.num; i++) {
+		if (get_key_translation(dstr, keycodes->list.array[i])) {
+			break;
+		}
+	}
+	*/
```
I'm not sure what were you intending to test here. Was it multi-layout?

kkartaltepe replied:
Oh sorry this was copied from the X11 impl, meant to delete.

GeorgesStavracas commented on libobs/obs-nix-wayland.c:
```diff
@@ -51,24 +281,1337 @@ obs_nix_wayland_hotkeys_platform_is_pressed(obs_hotkeys_platform_t *context,
+	*/
+	obs_hotkeys_platform_t *plat = obs->hotkeys.platform_context;
+	// Translate the obs key back down to shift level 1 and then back to obs key.
+	xkb_keysym_t base_sym = plat->key_to_sym[plat->obs_to_key[key]][0];
```
Nitpick: could you please put `plat->obs_to_key[key]` in a separate variable?

GeorgesStavracas reviewed (changes requested):
I left some comments and questions, but overall this looks pretty good to me. I didn't think of doing a roundtrip to get the wl_keyboard, what I had in mind is somehow get that from Qt, but that's probably a bad idea. What's here is an interesting approach.

kkartaltepe commented on libobs/obs-nix-wayland.c:
```diff
@@ -19,7 +19,224 @@
+		return;
+
+	const xkb_keysym_t *syms;
+	// My setup doesnt have multiple layouts, do people actually use these?
+	// int num_layouts = xkb_keymap_num_layouts_for_key(keymap, key);
+	// blog(LOG_INFO, "key layouts: %d", num_layouts);
```
based on https://www.x.org/releases/X11R7.6/doc/xorg-docs/input/XKB-Config.html#id2520758 and an overview of the keyboard layouts preloaded on archlinux. There are no multi-layout keyboards defined by default, the only case for this would be users with custom defined xkb keymaps (so like 3 people?) where there appears to be like 4 possible layout levels.

I dont mind also hardcoding in layouts if we assume there can be up to 4 of them, though im not sure we really want to support arbitrary groups too, especially not if IMEs do change keymaps instead.

kkartaltepe replied:
I played with some IME's and gnome, it looks like IMEs will use their own path for inputting characters for text input and do not use this mechanism at all.

For users configuring keyboards via gnome settings and not using IMEs gnome will only ever configure 4 layouts/groups at a time to remain X11 compatible. If you switch past the 4th keyboard gnome will reconfigure the wayland keyboard.

In my minimal playing this code seems to work fine for multiple languages once we handle multiple layouts/groups.

GeorgesStavracas replied:
Alright, thanks for checking this

kkartaltepe replied:
Pushed changes, I build the mapping tables any time users switch layouts. In cases where a user has something like a qwerty and azerty layout, scancode for AD01 might be bound to 'q' and 'a' which isnt quite correct. Switching keymaps takes 0.016ms on my machine so I think there is no problem with computing the lookup tables when users swap layouts to make the code simpler.

GeorgesStavracas reviewed (changes requested):
The code looks good to me, I think it's good to go. However, I'd like to ask capitalizing `wayland` → `Wayland`

kkartaltepe commented:
Updated, added some version checks when blinding `wl_seat`, checking the definition nothing important has changed for the past 8 years since v4 so we will bind whatever the compositor reports up to v7.

I think this is good to merge now.
